src/main/cpp/src/datetime_parser.cu

```diff
@@ -0,0 +1,453 @@
+          current_segment_digits = 0;
+          i += 1;
+          tz = cudf::string_view(bytes + j, (bytes_length - j));
+          j = bytes_length - 1;
+        }
+        if (i == 6 && '.' != b) {
+          i += 1;
+        }
+      } else {
+        if (i < segments_len && (':' == b || ' ' == b)) {
+          if (!is_valid_digits(i, current_segment_digits)) {
+            return thrust::make_pair(error_compoments, error_time_zone);
+          }
+          segments[i] = current_segment_value;
+          current_segment_value = 0;
+          current_segment_digits = 0;
+          i += 1;
+        } else {
+          return thrust::make_pair(error_compoments, error_time_zone);
+        }
+      }
+    } else {
+      if (6 == i) {
+        digits_milli += 1;
+      }
+      // We will truncate the nanosecond part if there are more than 6 digits, which results
+      // in loss of precision
+      if (6 != i || current_segment_digits < 6) {
+        current_segment_value = current_segment_value * 10 + parsed_value;
+      }
+      current_segment_digits += 1;
+    }
+    j += 1;
+  }
+
+  if (!is_valid_digits(i, current_segment_digits)) {
+    return thrust::make_pair(error_compoments, error_time_zone);
+  }
+  segments[i] = current_segment_value;
+
+  while (digits_milli < 6) {
+    segments[6] *= 10;
+    digits_milli += 1;
+  }
+
+  cudf::string_view timze_zone;
+  if (tz.has_value()) {
+    timze_zone = format_zone_id(tz.value());
+  } else {
+    timze_zone = default_time_zone;
+  }
+
+  segments[0] *= year_sign.value_or(1);
+  // above is translated from Spark.
+
+  // set components
+  auto components = timestamp_components{segments[0],
+                                         static_cast<int8_t>(segments[1]),
+                                         static_cast<int8_t>(segments[2]),
+                                         static_cast<int8_t>(segments[3]),
+                                         static_cast<int8_t>(segments[4]),
+                                         static_cast<int8_t>(segments[5]),
+                                         segments[6]};
+  if (just_time) {
+    components.year = components.month = components.day = -1;
+  }
+  return thrust::make_pair(components, timze_zone);
+}
+
+struct parse_timestamp_string_fn {
+  cudf::column_device_view const d_strings;
+  cudf::string_view default_time_zone;
+
+  __device__ cudf::timestamp_us operator()(const cudf::size_type &idx) const {
+    auto const d_str = d_strings.element<cudf::string_view>(idx);
+    auto components_tz = parse_string_to_timestamp_components_tz(d_str, default_time_zone);
+    return create_timestamp_from_components_and_zone(components_tz.first, components_tz.second);
+  }
+};
+
+/**
+ *
+ * Trims and parses timestamp string column to a timestamp column and a time zone
+ * column
+ *
+ */
+std::unique_ptr<cudf::column> parse_string_to_timestamp_and_time_zone(
+    cudf::strings_column_view const &input, cudf::string_view default_time_zone,
```

Comment:
I am not sure, but I feel it might be better to just pass the timezone as a simple (host-side) string since it should be a very short string.  Then, there are a lot of approaches to handle it.
For instance,
1. create a single-size `column_device_view` 
2. wrap it with `thrust::make_constant_iterator`
3. bind to `input` with `thrust::make_zip_iterator`

Reply:
Done